File: bmlib/db/operations.py
# bmlib — shared library for biomedical literature tools
# Copyright (C) 2024-2026 Dr Horst Herb
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

"""Pure-function query helpers.

All functions take a DB-API connection as their first argument.  SQL is
passed in directly — callers are responsible for writing backend-appropriate
SQL (``?`` for SQLite, ``%s`` for PostgreSQL).

These helpers wrap the DB-API cursor pattern to provide a cleaner call
interface while remaining completely transparent.
"""

from __future__ import annotations

import logging
from collections.abc import Sequence
from typing import Any

from bmlib.db.backend import is_sqlite
from bmlib.db.transactions import owns_commit

logger = logging.getLogger(__name__)


def execute(conn: Any, sql: str, params: Sequence = ()) -> Any:
    """Execute a single statement and return the cursor.

    Useful for INSERT / UPDATE / DELETE where you might need
    ``cursor.lastrowid`` or ``cursor.rowcount``.
    """
    cur = conn.cursor()
    cur.execute(sql, params)
    return cur


def executemany(conn: Any, sql: str, params_seq: Sequence[Sequence]) -> None:
    """Execute a statement for each parameter set in *params_seq*."""
    cur = conn.cursor()
    cur.executemany(sql, params_seq)


def fetch_one(conn: Any, sql: str, params: Sequence = ()) -> Any:
    """Execute and return the first row, or ``None``."""
    cur = conn.cursor()
    cur.execute(sql, params)
    return cur.fetchone()


def fetch_all(conn: Any, sql: str, params: Sequence = ()) -> list[Any]:
    """Execute and return all rows."""
    cur = conn.cursor()
    cur.execute(sql, params)
    return cur.fetchall()


def fetch_scalar(conn: Any, sql: str, params: Sequence = ()) -> Any:
    """Execute and return the first column of the first row, or ``None``."""
    row = fetch_one(conn, sql, params)
    if row is None:
        return None
    # psycopg2's RealDictRow is a dict subclass keyed by column name, so
    # ``row[0]`` raises KeyError there and "first column" has to mean the first
    # *value*. sqlite3.Row is not a dict (it has keys() but no values()), so it
    # keeps the index path.
    if isinstance(row, dict):
        values = list(row.values())
        return values[0] if values else None
    try:
        return row[0]
    except (IndexError, KeyError):
        return None


def table_exists(conn: Any, name: str) -> bool:
    """Check whether a table exists (works on both SQLite and PostgreSQL)."""
    if is_sqlite(conn):
        row = fetch_one(
            conn,
            "SELECT 1 FROM sqlite_master WHERE type='table' AND name=?",
            (name,),
        )
    else:
        row = fetch_one(
            conn,
            "SELECT 1 FROM information_schema.tables WHERE table_name=%s",
            (name,),
        )
    return row is not None


def _split_sql_statements(script: str) -> list[str]:
    """Split a multi-statement SQL script into individual statements.

    Splits on semicolons that are outside string literals, comments, and
    compound statement bodies. Handles single/double-quoted strings (with SQL
    ``''`` escaping), ``--`` line comments, and ``/* */`` block comments.

    Trigger bodies are handled: a ``CREATE TRIGGER ... BEGIN ... END;`` block
    contains semicolons that terminate the body's statements, not the
    ``CREATE TRIGGER`` itself, so splitting on them would hand SQLite a
    fragment and raise "incomplete input". Nesting is tracked by counting
    ``BEGIN``/``CASE`` against ``END``; counting ``CASE`` matters because a
    ``CASE ... END`` expression inside a body would otherwise close it early.
    A bare ``BEGIN`` (transaction control) does not open a body — the counter
    only engages once ``TRIGGER`` has been seen in the current statement.
    """
    statements: list[str] = []
    buf: list[str] = []
    i = 0
    n = len(script)
    quote: str | None = None
    depth = 0  # BEGIN/CASE nesting within a compound statement
    in_trigger = False  # current statement is a CREATE TRIGGER
    word: list[str] = []

    def flush_word() -> None:
        """Consume the just-scanned word and update nesting state."""
        nonlocal depth, in_trigger
        if not word:
            return
        w = "".join(word).upper()
        word.clear()
        if w == "TRIGGER" and depth == 0:
            in_trigger = True
        elif w in ("BEGIN", "CASE"):
            if in_trigger:
                depth += 1
        elif w == "END" and depth > 0:
            depth -= 1

    while i < n:
        ch = script[i]
        if quote:
            buf.append(ch)
            if ch == quote:
                # A doubled quote is an escaped quote, not a terminator.
                if i + 1 < n and script[i + 1] == quote:
                    buf.append(script[i + 1])
                    i += 2
                    continue
                quote = None
            i += 1
            continue
        # Not inside a string literal.
        if ch.isalnum() or ch == "_":
            word.append(ch)
            buf.append(ch)
            i += 1
            continue
        # Any other character ends the word currently being scanned.
        flush_word()
        if ch in ("'", '"'):
            quote = ch
            buf.append(ch)
            i += 1
        elif ch == "-" and i + 1 < n and script[i + 1] == "-":
            # Line comment: skip to end of line.
            j = script.find("\n", i)
            i = n if j == -1 else j
        elif ch == "/" and i + 1 < n and script[i + 1] == "*":
            # Block comment: skip to closing */.
            j = script.find("*/", i + 2)
            i = n if j == -1 else j + 2
        elif ch == ";" and depth == 0:
            stmt = "".join(buf).strip()
            if stmt:
                statements.append(stmt)
            buf = []
            in_trigger = False
            i += 1
        else:
            buf.append(ch)
            i += 1
    flush_word()
    tail = "".join(buf).strip()
    if tail:
        statements.append(tail)
    return statements


def create_tables(conn: Any, schema_sql: str) -> None:
    """Execute a (possibly multi-statement) schema DDL string.

    Statements are executed individually via a cursor for both backends.
    SQLite's ``executescript()`` is deliberately avoided because it issues an
    implicit ``COMMIT`` before running, which would break the atomicity of a
    surrounding :func:`~bmlib.db.transactions.transaction` (e.g. during
    migrations). Executing statements one at a time keeps the DDL inside the
    active transaction — SQLite supports transactional DDL — so a failure
    rolls back cleanly.

    Either backend skips the commit when called from inside a
    :func:`~bmlib.db.transactions.transaction` block, leaving the DDL atomic
    with the rest of that block (e.g. the migration runner's). PostgreSQL
    supports transactional DDL too.
    """
    cur = conn.cursor()
    if is_sqlite(conn):
        for stmt in _split_sql_statements(schema_sql):
            cur.execute(stmt)
        # Persist when called standalone, but leave commit to the caller when
        # an explicit transaction is active (e.g. the migration runner), so
        # the DDL stays atomic with the rest of that transaction.
        if not conn.in_transaction:
            conn.commit()
    else:
        cur.execute(schema_sql)
        # psycopg2 counts a plain SELECT as opening a transaction, so the
        # driver's status cannot answer "am I inside someone's block?" — ask
        # bmlib's own nesting count instead.
        if owns_commit(conn):
            conn.commit()

File: spikes/db-rs/src/operations.rs
// bmlib — shared library for biomedical literature tools
// Copyright (C) 2024-2026 Dr Horst Herb
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

//! Pure-function query helpers.
//!
//! Same shape as the Python module: every function takes the connection as its
//! first argument and is transparent about the SQL. The one difference is that
//! callers always write `?`; the PostgreSQL numbered form is produced here,
//! not at the call site.

use crate::backend::{adapt_sql, Dialect};
use crate::db::Db;
use crate::error::Result;
use crate::split::split_sql_statements;
use crate::value::{Row, Value};

/// Run a statement, returning the number of rows affected.
pub fn execute(db: &mut dyn Db, sql: &str, params: &[Value]) -> Result<u64> {
    let sql = adapt_sql(sql, db.dialect());
    db.execute_raw(&sql, params)
}

/// Run a statement once per parameter set.
pub fn executemany(db: &mut dyn Db, sql: &str, rows: &[Vec<Value>]) -> Result<()> {
    let sql = adapt_sql(sql, db.dialect());
    for params in rows {
        db.execute_raw(&sql, params)?;
    }
    Ok(())
}

/// Run a query and return every row.
pub fn fetch_all(db: &mut dyn Db, sql: &str, params: &[Value]) -> Result<Vec<Row>> {
    let sql = adapt_sql(sql, db.dialect());
    db.query_raw(&sql, params)
}

/// Run a query and return the first row, if any.
pub fn fetch_one(db: &mut dyn Db, sql: &str, params: &[Value]) -> Result<Option<Row>> {
    Ok(fetch_all(db, sql, params)?.into_iter().next())
}

/// Run a query and return the first column of the first row, if any.
///
/// Python needs a backend branch here, because `sqlite3.Row` indexes by
/// position while psycopg2's `RealDictRow` is a dict that raises `KeyError` on
/// `row[0]`, so "first column" means two different lookups. [`Row`] keeps
/// values in column order on both backends, so the branch is gone.
pub fn fetch_scalar(db: &mut dyn Db, sql: &str, params: &[Value]) -> Result<Option<Value>> {
    Ok(fetch_one(db, sql, params)?.and_then(|r| r.at(0).cloned()))
}

/// Whether a table exists, on either backend.
pub fn table_exists(db: &mut dyn Db, name: &str) -> Result<bool> {
    let sql = match db.dialect() {
        Dialect::Sqlite => "SELECT 1 FROM sqlite_master WHERE type='table' AND name=?",
        Dialect::Postgres => "SELECT 1 FROM information_schema.tables WHERE table_name=?",
    };
    Ok(fetch_one(db, sql, &[Value::Text(name.to_string())])?.is_some())
}

/// Execute a (possibly multi-statement) schema DDL string.
///
/// Statements are run one at a time. `rusqlite`'s `execute_batch` is avoided
/// for the same reason the Python port avoids `executescript()` — a batch
/// primitive that commits behind your back would break the atomicity of a
/// surrounding [`crate::transaction`]. Both backends support transactional
/// DDL, so a failure inside a block rolls the DDL back with everything else.
///
/// Unlike the Python version this needs no commit of its own: a connection is
/// in autocommit mode, and inside a block the opener owns the commit. That is
/// what retires the `owns_commit()` call the Python version had to make.
///
/// **Divergence from the Python port, deliberate:** Python splits for SQLite
/// only and hands psycopg2 the whole script, because psycopg2 accepts one. The
/// split is correct on both backends, so doing it unconditionally removes a
/// dialect branch — at the cost of one round trip per statement, which for
/// `publications/schema.py` is 25. Two things to check before adopting this in
/// earnest: measure that cost against a real server, and note that
/// [`crate::split::split_sql_statements`] does not know PostgreSQL's
/// dollar-quoted bodies (`$$ ... $$`), which bmlib's schema does not use but a
/// `CREATE FUNCTION` would.
pub fn create_tables(db: &mut dyn Db, schema_sql: &str) -> Result<()> {
    for stmt in split_sql_statements(schema_sql) {
        db.execute_raw(&stmt, &[])?;
    }
    Ok(())
}
